lib/Feature/FeatureModelWriter.cpp

```diff
@@ -0,0 +1,347 @@
+#include "vara/Feature/FeatureModelWriter.h"
+#include "vara/Feature/FeatureModel.h"
+#include <llvm/Support/Casting.h>
+
+#include "XmlConstants.h"
+
+#include "libxml/xmlwriter.h"
+
+#define CHECK_RC                                                               \
+  if (RC < 0) {                                                                \
+    return RC;                                                                 \
+  }
+
+namespace vara::feature {
+
+static constexpr xmlChar INDENTATIONSTRING[] = "  ";
+static constexpr char ENCODING[] = "UTF-8";
+
+int FeatureModelXmlWriter::writeFeatureModel(std::string Path) {
+  int RC;
+  std::unique_ptr<xmlTextWriter, void (*)(xmlTextWriterPtr)> Writer(
+      xmlNewTextWriterFilename(Path.data(), 0), &xmlFreeTextWriter);
+  if (!Writer) {
+    return false;
+  }
+
+  RC = writeFeatureModel(Writer.get());
+  return RC;
+}
+
+std::optional<std::string> FeatureModelXmlWriter::writeFeatureModel() {
+  int RC;
+
+  std::unique_ptr<xmlDocPtr, void (*)(xmlDocPtr *)> DocPtrPtr(
+      new xmlDocPtr(), [](xmlDocPtr *Ptr) {
+        xmlFreeDoc(*Ptr);
+        delete (Ptr);
```

Comment:
Why do you need to call delete on the ptr after using the xmlfree function? This looks like a double free to me.

Reply:
xmlNewTextWriterDoc needs a xmlDocPtr *. Before I declared a xmlDocPtr and passed a reference to that, but that didn't work with the unique_ptr.
So, the unique pointer holds xmlDocPtr, which I need to create (here with `new xmlDocPtr()`). The xmlDoc is freed by `xmlFreeDoc` and the xmlDocPtr is freed by the delete. 

If there is a way to get a xmlDocPtr * out of a unique_ptr<xmlDoc>, it would make this way easier, but I believe that doesn't work.

I also just noticed that this might call `xmlFreeDoc` with an uninitialized pointer, so I will fix it, if the rest is an acceptable solution.